rafthttp/stream.go

```diff
@@ -392,45 +392,61 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
 	lv := semver.Must(semver.NewVersion(version.Version))
 	if compareMajorMinorVersion(rv, lv) == -1 && !checkStreamSupport(rv, t) {
 		resp.Body.Close()
+		cr.picker.unreachable(u)
 		return nil, errUnsupportedStreamType
 	}
 
-	switch resp.StatusCode {
-	case http.StatusGone:
+	err = cr.checkStreamResponse(resp)
+	if err != nil {
 		resp.Body.Close()
-		err := fmt.Errorf("the member has been permanently removed from the cluster")
-		select {
-		case cr.errorc <- err:
-		default:
+		cr.picker.unreachable(u)
+		// errMemberRemoved is a critical error since a removed member should
+		// always be stopped. So we use reportCriticalError to report it to errorc.
+		if err == errMemberRemoved {
+			reportCriticalError(err, cr.errorc)
 		}
 		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// checkStreamResponse checks the response of one stream request.
+// It returns nil if the response indicates that stream request succeeds,
+// or it returns error.
+// If the response indicates stream request failure, the function may read the
+// response body for verbosal failure message.
+func (cr *streamReader) checkStreamResponse(resp *http.Response) error {
```

Comment:
This is ugly. We should not repeat `Stream` since this is already a streamReader. I think I have suggested something else.